pkg/updater/api.go

```diff
@@ -0,0 +1,149 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package updater
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/gorilla/mux"
+)
+
+type startExperimentRequest struct {
+	Version string `json:"version"`
+}
+
+type startExperimentResponse struct {
+	Error string `json:"error,omitempty"`
+}
+
+type stopExperimentRequest struct {
+}
+
+type stopExperimentResponse struct {
+	Error string `json:"error,omitempty"`
+}
+
+type promoteExperimentRequest struct {
+}
+
+type promoteExperimentResponse struct {
+	Error string `json:"error,omitempty"`
+}
+
+// LocalAPI is a locally exposed API to interact with the updater.
+type LocalAPI struct {
+	updater  *Updater
+	listener net.Listener
+}
+
+// NewLocalAPI returns a new LocalAPI.
+func NewLocalAPI(updater *Updater) (*LocalAPI, error) {
+	socketPath := path.Join(updater.repositoryPath, fmt.Sprintf("%s-updater.sock", updater.pkg))
+	err := os.RemoveAll(socketPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not remove socket: %w", err)
+	}
+	listener, err := net.Listen("unix", socketPath)
+	if err != nil {
+		return nil, err
+	}
+	if err := os.Chmod(socketPath, 0700); err != nil {
```

Comment:
In general I think it's a better a simpler approach than what we have in the agent since it offers strong security and does not require yet another port to be reserved. Windows might be a problem in the long run as it's only supported on Win10+ IIRC but I'd still want to start with this simpler approach first as it's not clear yet if we want this local API or not in the long run.

Reply:
That's fair and yeah the Windows support would be the main problem coming to mind. With sockets, there may be race conditions affecting the parent folder permissions that might need to be considered though.

Reply:
You're right that we have to be careful about the parent directory. Here given that the parent repository is owned by the updater (central place to store packages), I believe we are in the green. It's a bit tricky to fully test right now since this directory will be created at the updater install. I'll make a note that we need to be careful about permissions there, not only for the socket but also because we don't want the world to write to the packages themselves obviously 😄